### ISSUE TITLE
Considerations for worker metrics in dashboard / performance reports

### ISSUE THREAD

**Issue by charlesbluca** [dask/distributed#4668]
Motivated by dask/distributed#4494, I've been thinking a lot about how worker metrics (including GPU memory/utilization) could be represented in the Distributed dashboard / performance reports; this has culminated in a bunch of PRs playing around with dashboard components and the system monitor on my end, but I realize that since most of that work has only included the same ~5 people, it could be good to open wider discussion on this.

Some of the information desired from #4494 is:

- information on a worker metric over the course of an operation
- peak/average value of a metric during an operation

Some questions/thoughts that come to mind:

- *Would time series plots be useful for tracking worker metrics?* My first thought was to have something like the dashboard's System tab for workers, but it would require some progress in Bokeh (bokeh/bokeh#11101) to get working.
- *Where could statistics (mean/max) of a metric be displayed?* My first thought was either as a label / hover tool in a plot of the metric, or in a sortable table the the one currently used in the Workers tab.
- *Should GPU metrics stay separate from other metrics?* Currently, GPU information is only available in the standalone GPU tab added by dask/distributed#4556, but it shouldn't be difficult to add it to something like the Workers table or a hypothetical Workers time series tab. One thing to note here is that since GPU info is collected and displayed conditional on a user having `pynvml` installed, GPU info will show up even if a user isn't working with GPUs.

Interested to hear more opinions on this.

**Comment by mrocklin** [comment]
I think that tracking aggregate worker metrics over time would be a great idea.  Probably we shouldn't include a line per worker, which might get messy (we used to do this and it got messy).  It also wouldn't scale well.

In general I would like for us to think about scalable visualizations in the future.  I think that this is an interesting area for us to innovate.

In the case of per-worker timeseries plots I recommend that we display a distribution of {min, -1sigma, mean, -2sigma, max} or something similar.  I suspect that there are charts/glyphs like stacked plots that would work well here.  This would, I think, convey what we want conveyed (things are tight around 60% except for some abberational worker), be pretty, and work well at scale.

**Comment by quasiben** [comment]
Maybe this is where a box and whisker plot comes in handy.  My understanding of these kinds of plot is that it allows us to see these kinds of summary statistics vertically for each worker (min/max/quantile) and then we would have N number of boxplots horizontally for the number of workers.  This may be somewhat scalable but would break down at around 1000+ workers ?  I'm not sure -- or maybe because there isn't any inherit ordering of the workers we could order by Max or let the user choose the ordering.

One thing to note about GPU metrics in particular is that often Dask GPU workloads will pre-allocate most GPU memory so won't really really see much variation here.  This is not true for CPU workloads where it would be very helpful seeing a spike in memory consumption

**Comment by mrocklin** [comment]
Yeah, I think that a plot that had 1000 whisker plots would not be very
informative.  I would recommend aggregating across workers and seeing what
that felt like.

**Comment by charlesbluca** [comment]
> In the case of per-worker timeseries plots I recommend that we display a distribution of {min, -1sigma, mean, -2sigma, max} or something similar.

This sounds like a good idea - I don't know if Bokeh explicitly has stacked plots, but [band annotations](https://docs.bokeh.org/en/latest/docs/user_guide/annotations.html#bands) seem similar.

If we show the aggregations in a time series, I feel like there's the potential for min and max to reach some outlier and stay there for the remainder of a run, which might not be as helpful as mean/quartile (which I'm assuming would be changing somewhat frequently).

> Maybe this is where a box and whisker plot comes in handy. My understanding of these kinds of plot is that it allows us to see these kinds of summary statistics vertically for each worker (min/max/quantile) and then we would have N number of boxplots horizontally for the number of workers.

I like this idea too, and think that band annotations could also be useful here if we wanted to visualize the global mean/quartile across all workers. I don't know if this would be cumbersome, but I was also imagining something like this coupled with a slider widget so that users could pan through the worker stats at different points in time.

I'm wondering if, when aggregating metrics, if it could be useful to have some additional context, maybe in a hover tool - for example, we could keep track of what task was running when a metric reached its peak.

**Comment by charlesbluca** [comment]
With talk of plot scalability, I feel like it could be interesting to introduce components that are conditional on worker count - for example, maybe a box-whisker plot could be useful up to some N workers, after which that plot is switched out for something further aggregated to take its place.

**Comment by mrocklin** [comment]
Yes, we do something similar to this today for the memory plots (also a per-worker plot, so not great).  They switch to histograms above 50 workers.  It works ok, although I think that users find it a little confusing.

**Comment by charlesbluca** [comment]
That sounds interesting - my thought was that we could make this transition seamless by using the same or similar plots, with elements added/removed when they lose functionality.

For example, for a worker time series we could have the worker lines in addition to min/max/quartile banding; when the worker count gets sufficiently high, we could remove the per-worker lines and leave only the banding, so users should already know what it represents. In a case like the per-worker memory plots, maybe having a box and whisker plot somewhat visible that becomes prominent when the worker count is high.

Do we know how other distributed handle this problem? I'm currently checking out Spark and Ray, but haven't seen too many plots that seem well suited for high worker count. Also cc @jacobtomlinson, who introduced me to Weave Scope.

**Comment by jacobtomlinson** [comment]
Generally the dashboards give us two things:

- **Reassurance that things are happening.** For this aesthetics can win over practicality. For example the task stream is my favourite plot, but the information could be displayed in a more boring but useful way. For example [this plot](https://github.com/dask/distributed/issues/4260#issuecomment-738183697) may be a better replacement, but is less visually exciting. But I enjoy running code and seeing it distribute on the cluster and the pleasant design gets newcomes interested. The cluster map plot is similar.
- **Answering specific questions about the cluster**. Am I running out of memory? Which parts of my code took the longest to run? Am I spending lots of time spilling to disk? Are my workers compute, memory or IO bound?

Assuming that this discussion is about the latter it may be helpful in this discussion to try and consider that questions users will be trying to answer with the dashboard. Specifically what questions can they not easily answer today. Then with those questions try to work backwards to the plots that we need to create to enable them to be answered.

Some examples:
- What percentage of time do my workers spend on memory transfer?
- I have 1000 workers, are they clustered into different bottlenecks (compute, memory, io)?
